hal/mbed_sleep_manager.c

```diff
@@ -19,13 +19,116 @@
 #include "sleep_api.h"
 #include "mbed_error.h"
 #include <limits.h>
+#include <stdio.h>
 
 #if DEVICE_SLEEP
 
 // deep sleep locking counter. A target is allowed to deep sleep if counter == 0
 static uint16_t deep_sleep_lock = 0U;
 
-void sleep_manager_lock_deep_sleep(void)
+#ifdef MBED_SLEEP_TRACING_ENABLED
+
+// Length of the identifier extracted from the driver name to store for logging.
+#define IDENTIFIER_WIDTH 7
+// Number of drivers that can be stored in the structure
+#define STATISTIC_COUNT  10
+
+typedef struct sleep_statistic {
+    char identifier[IDENTIFIER_WIDTH];
+    uint8_t count;
+} sleep_statistic_t;
+
+static sleep_statistic_t sleep_stats[STATISTIC_COUNT];
+
+static const char* strip_path(const char* const filename)
+{
+    char *output = strrchr(filename, '/');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    output = strrchr(filename, '\\');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    return filename;
+}
+
+static size_t sleep_tracker_find_index(const char *const filename)
+{
+    char temp[IDENTIFIER_WIDTH];
+    strncpy(temp, filename, IDENTIFIER_WIDTH);
+    temp[IDENTIFIER_WIDTH - 1] = '\0';
+
+    // Search for the a driver matching the current name and return it's index
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (strcmp(sleep_stats[i].identifier, temp) == 0) {
+            return i;
+        }
+    }
+
+    // If no driver was found currently in the structure, find the first array
+    // index that hasn't been used by a driver and fill it, then return the
+    // index.
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].identifier[0] == '\0') {
+            core_util_critical_section_enter();
+            strncpy(sleep_stats[i].identifier, temp, sizeof(temp));
+            core_util_critical_section_exit();
+
+            return i;
+        }
+    }
+
+    // Panic if there are no free indexes left to track with
+    error("No free indexes left to use in mbed stats tracker");
+
+    return -1;
+}
+
+static void sleep_tracker_print_stats(void)
+{
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].count == 0) {
+            continue;
+        }
+
+        if (sleep_stats[i].identifier[0] == '\0') {
+            return;
+        }
+
+        printf("[id: %s, count: %u]\r\n", sleep_stats[i].identifier,
+                                          sleep_stats[i].count);
+    }
+}
+
+void sleep_tracker_lock(const char* const filename, int line)
```

Comment:
It seems like that would require changing the user-facing API, and requiring anyone who uses the sleep tracker to keep track of the index. Currently, the sleep tracing is transparent to the user. I don't think that requiring them to track the index is worth the benefit of not having to look up the index during unlock, especially when this is meant intended primarily as an optional debugging feature.

Reply:
Worth adding that sleep tracing is only available in debug mode and just the fact of using it will prevent the board from sleeping, so it's purely development time debug feature.

Reply:
> Worth adding that sleep tracing is only available in debug mode and just the fact of using it will prevent the board from sleeping, so it's purely development time debug feature

I dont see in this change set where that would be the case. Can you point me at it?

Reply:
Implicitly by using UART.

Reply:
I dont understand how that is possible given UART is available in develop and release profiles as well.